File: rle.js
// rle.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const process = require('process');

const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const ESCAPE = '\\';

function compress(text) {
    if (!text) return '';
    const result = [];
    let i = 0, n = text.length;
    while (i < n) {
        const ch = text[i];
        let j = i + 1;
        while (j < n && text[j] === ch) j++;
        const count = j - i;
        if (count >= 3) {
            result.push(ch, ESCAPE, count.toString());
        } else {
            for (let k = 0; k < count; k++) {
                if (ch === ESCAPE) result.push(ESCAPE);
                result.push(ch);
            }
        }
        i = j;
    }
    return result.join('');
}

function decompress(text) {
    if (!text) return '';
    const result = [];
    let i = 0, n = text.length;
    while (i < n) {
        const ch = text[i];
        if (ch === ESCAPE) {
            if (i + 1 < n && text[i + 1] === ESCAPE) {
                result.push(ESCAPE);
                i += 2;
                continue;
            }
            if (i + 1 >= n) throw new Error('Unexpected end after escape');
            const repeatChar = text[i + 1];
            i += 2;
            let numStr = '';
            while (i < n && text[i] >= '0' && text[i] <= '9') {
                numStr += text[i];
                i++;
            }
            if (!numStr) throw new Error('Missing number after escape');
            const count = parseInt(numStr, 10);
            for (let k = 0; k < count; k++) result.push(repeatChar);
        } else {
            result.push(ch);
            i++;
        }
    }
    return result.join('');
}

function readInput(filename) {
    if (!filename || filename === '-') {
        return fs.readFileSync(0, 'utf-8');
    }
    return fs.readFileSync(filename, 'utf-8');
}

function writeOutput(filename, content) {
    if (!filename || filename === '-') {
        process.stdout.write(content);
    } else {
        fs.writeFileSync(filename, content, 'utf-8');
    }
}

function parseArgs() {
    const args = process.argv.slice(2);
    const opts = { mode: '', input: '', output: '', verbose: false };
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '-v' || args[i] === '--verbose') {
            opts.verbose = true;
        } else if (!opts.mode && (args[i] === 'compress' || args[i] === 'decompress')) {
            opts.mode = args[i];
        } else if (!opts.input) {
            opts.input = args[i];
        } else {
            opts.output = args[i];
        }
    }
    return opts;
}

function main() {
    const opts = parseArgs();
    if (!opts.mode) {
        console.log(colorize('Usage: node rle.js compress|decompress [options] <input> [output]', 'yellow'));
        console.log('Options: -v, --verbose  show statistics');
        process.exit(1);
    }

    let data;
    try {
        data = readInput(opts.input);
    } catch (err) {
        console.log(colorize('Error reading input: ' + err.message, 'red'));
        process.exit(1);
    }
    const origSize = Buffer.byteLength(data, 'utf8');

    let result;
    try {
        if (opts.mode === 'compress') result = compress(data);
        else result = decompress(data);
    } catch (err) {
        console.log(colorize('Error: ' + err.message, 'red'));
        process.exit(1);
    }
    const compSize = Buffer.byteLength(result, 'utf8');

    if (opts.verbose && opts.mode === 'compress') {
        const ratio = origSize ? compSize / origSize : 1;
        const saving = (1 - ratio) * 100;
        console.log(colorize(`Original size: ${origSize} bytes`, 'yellow'));
        console.log(colorize(`Compressed size: ${compSize} bytes`, 'yellow'));
        console.log(colorize(`Ratio: ${ratio.toFixed(2)}x`, 'green'));
        console.log(colorize(`Space saving: ${saving.toFixed(1)}%`, 'green'));
    }

    try {
        writeOutput(opts.output, result);
        if (opts.output && opts.output !== '-') {
            console.log(colorize('Result written to ' + opts.output, 'green'));
        }
    } catch (err) {
        console.log(colorize('Error writing output: ' + err.message, 'red'));
        process.exit(1);
    }
}

main();

File: rle.py
# rle.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import argparse
import os

# ANSI colors
COLORS = {
    'green': '\033[92m',
    'red': '\033[91m',
    'yellow': '\033[93m',
    'reset': '\033[0m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

ESCAPE = '\\'

def compress(text):
    """Сжатие RLE с поддержкой Unicode (символы)."""
    if not text:
        return ''
    result = []
    i = 0
    n = len(text)
    while i < n:
        ch = text[i]
        j = i + 1
        while j < n and text[j] == ch:
            j += 1
        count = j - i
        if count >= 3:
            result.append(ch)
            result.append(ESCAPE)
            result.append(str(count))
        else:
            for _ in range(count):
                if ch == ESCAPE:
                    result.append(ESCAPE)
                result.append(ch)
        i = j
    return ''.join(result)

def decompress(text):
    """Распаковка RLE."""
    if not text:
        return ''
    result = []
    i = 0
    n = len(text)
    while i < n:
        ch = text[i]
        if ch == ESCAPE:
            if i + 1 < n and text[i + 1] == ESCAPE:
                result.append(ESCAPE)
                i += 2
                continue
            if i + 1 >= n:
                raise ValueError("Unexpected end after escape")
            char = text[i + 1]
            i += 2
            num_str = ''
            while i < n and text[i].isdigit():
                num_str += text[i]
                i += 1
            if not num_str:
                raise ValueError("Missing number after escape")
            count = int(num_str)
            result.append(char * count)
        else:
            result.append(ch)
            i += 1
    return ''.join(result)

def read_input(filename):
    if not filename or filename == '-':
        return sys.stdin.read()
    with open(filename, 'r', encoding='utf-8') as f:
        return f.read()

def write_output(filename, content):
    if not filename or filename == '-':
        sys.stdout.write(content)
    else:
        with open(filename, 'w', encoding='utf-8') as f:
            f.write(content)

def main():
    parser = argparse.ArgumentParser(description="RLE Advanced Compressor/Decompressor")
    parser.add_argument('mode', choices=['compress', 'decompress'], help='Режим работы')
    parser.add_argument('input', nargs='?', help='Входной файл (или - для stdin)')
    parser.add_argument('output', nargs='?', help='Выходной файл (или - для stdout)')
    parser.add_argument('-v', '--verbose', action='store_true', help='Показать статистику сжатия')
    args = parser.parse_args()

    try:
        data = read_input(args.input)
    except Exception as e:
        sys.exit(colorize(f"Error reading input: {e}", 'red'))

    original_size = len(data.encode('utf-8'))

    try:
        if args.mode == 'compress':
            result = compress(data)
        else:
            result = decompress(data)
    except Exception as e:
        sys.exit(colorize(f"Error: {e}", 'red'))

    compressed_size = len(result.encode('utf-8'))

    if args.verbose and args.mode == 'compress':
        ratio = compressed_size / original_size if original_size > 0 else 1
        saving = (1 - ratio) * 100
        print(colorize(f"Original size: {original_size} bytes", 'yellow'))
        print(colorize(f"Compressed size: {compressed_size} bytes", 'yellow'))
        print(colorize(f"Ratio: {ratio:.2f}x", 'green'))
        print(colorize(f"Space saving: {saving:.1f}%", 'green'))

    try:
        write_output(args.output, result)
        if args.output and args.output != '-':
            print(colorize(f"Result written to {args.output}", 'green'))
    except Exception as e:
        sys.exit(colorize(f"Error writing output: {e}", 'red'))

if __name__ == '__main__':
    main()
